pkg/tools/etcd_tools.go

```diff
@@ -275,12 +287,50 @@ func (h *EtcdHelper) CreateObj(key string, obj runtime.Object, ttl uint64) error
 	return err
 }
 
+// Create adds a new object at a key unless it already exists. 'ttl' is time-to-live in seconds,
+// and 0 means forever. If no error is returned, out will be set to the read value from etcd.
+func (h *EtcdHelper) Create(key string, obj, out runtime.Object, ttl uint64) error {
+	data, err := h.Codec.Encode(obj)
+	if err != nil {
+		return err
+	}
+	if h.ResourceVersioner != nil {
+		if version, err := h.ResourceVersioner.ResourceVersion(obj); err == nil && version != 0 {
+			return errors.New("resourceVersion may not be set on objects to be created")
+		}
+	}
+	response, err := h.Client.Create(key, string(data), ttl)
+	if err != nil {
+		return err
+	}
+	if _, err := conversion.EnforcePtr(out); err != nil {
+		panic("unable to convert output object to pointer")
+	}
+	_, _, err = h.extractObj(response, err, out, false, false)
+	return err
+}
+
 // Delete removes the specified key.
 func (h *EtcdHelper) Delete(key string, recursive bool) error {
 	_, err := h.Client.Delete(key, recursive)
 	return err
 }
 
+// DeleteObj removes the specified key and returns the value that existed at that spot.
+func (h *EtcdHelper) DeleteObj(key string, out runtime.Object) error {
+	if _, err := conversion.EnforcePtr(out); err != nil {
+		panic("unable to convert output object to pointer")
+	}
+	response, err := h.Client.Delete(key, false)
```

Comment:
I'll spawn a follow up for that, it's possible via compareanddelete.

----- Original Message -----
> >  // Delete removes the specified key.
> >  func (h *EtcdHelper) Delete(key string, recursive bool) error {
> >  	_, err := h.Client.Delete(key, recursive)
> >  	return err
> >  }
> >  
> > +// DeleteObj removes the specified key and returns the value that existed
> > at that spot.
> > +func (h *EtcdHelper) DeleteObj(key string, out runtime.Object) error {
> > +	if _, err := conversion.EnforcePtr(out); err != nil {
> > +		panic("unable to convert output object to pointer")
> > +	}
> > +	response, err := h.Client.Delete(key, false)
> 
> Is it possible to apply a resourceVersion precondition on delete? That would
> be useful for delete-create races.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4419/files#r24681391